addons/l10n_it_edi_doi/models/declaration_of_intent.py

```diff
@@ -0,0 +1,323 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models, _
+from odoo.exceptions import UserError
+from odoo.tools.misc import formatLang
+
+
+class L10nItDeclarationOfIntent(models.Model):
+    _name = "l10n_it_edi_doi.declaration_of_intent"
+    _inherit = ['mail.thread.main.attachment', 'mail.activity.mixin']
+    _description = "Declaration of Intent"
+    _order = 'protocol_number_part1, protocol_number_part1'
+
+    state = fields.Selection([
+         ('draft', 'Draft'),
+         ('active', 'Active'),
+         ('revoked', 'Revoked'),
+         ('terminated', 'Terminated'),
+        ],
+        string="State",
+        tracking=True,
+        default='draft',
+        required=True,
+        readonly=True,
+        help="The state of this Declaration of Intent. \n"
+        "- 'Draft' means that the Declaration of Intent still needs to be confirmed before being usable. \n"
+        "- 'Active' means that the Declaration of Intent is usable. \n"
+        "- 'Terminated' designates that the Declaration of Intent has been marked as not to use anymore without invalidating usages of it. \n"
+        "- 'Revoked' means the Declaration of Intent should not have been used. You will probably need to revert previous usages of it, if any.\n")
+
+    company_id = fields.Many2one(
+        comodel_name='res.company',
+        string='Company',
+        index=True,
+        required=True,
+        default=lambda self: self.env.company._accessible_branches()[:1],
+    )
+
+    partner_id = fields.Many2one(
+        comodel_name='res.partner',
+        string='Partner',
+        index=True,
+        required=True,
+        domain=lambda self: ['|', ('is_company', '=', True), ('parent_id', '=', False)],
+    )
+
+    currency_id = fields.Many2one(
+        comodel_name='res.currency',
+        string='Currency',
+        default=lambda self: self.env.ref('base.EUR').id,
```

Comment:
I added the `raise_if_not_found=False`

I think having the field is useful so that the Monetary fields know the currency
and the currency symbol can be displayed in the form and tree views.

For now everything is in EUR since the declaration amount should be in EUR.
(Afaik) only Italian companies can get declarations. 
So we only sell to Italian companies anyways so the invoices should be in EUR.

I will recheck functionally about the currencies though.

Reply:
I checked: Everything in EUR is valid for now.
I edited my comment above.

Reply:
I don't see the point in storing a field that cannot be changed anyways, especially since we know (in the current state of the feature) that it's **always** going to be the company currency, and it currently only applies to Italian companies doing business with other Italian companies :sweat_smile:
(And you can still use the company currency in `formatLang()` for the views).
But I leave it up to you if you think it's better (or maybe I'm still missing something) :blush:

Reply:
Maybe there is a workaround, but if I just remove the `currency_id` field from the views there are the following problems:
- The Monetary fields are displayed without a currency sign
  (concerns i.e. DOI form / tree as well as account.move , SO tree view).
  They are potentially formatted differently too?
- The totals in the tree views of the Monetary fields are not even computed.
  (concerns i.e. account.move , SO tree view).

But we could take the currency from the company instead of "hardcoding" it to EUR too ofc?